### ISSUE TITLE
Support using Literal on __getattr__ and __getattribute__

### ISSUE THREAD

**Issue by linw1995** [python/mypy#8203]
Hi! Feature request for using `Literal` on `__getattr__` and `__getattribute__`.

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```python
from typing import Literal

LiteralStrHello = Literal["hello"]
LiteralStrWorld = Literal["world"]


class A:
    def __getattr__(self, name: LiteralStrHello) -> LiteralStrWorld:
        assert name == "hello"
        return "world"

    def __getattribute__(self, name: LiteralStrHello) -> LiteralStrWorld:
        assert name == "hello"
        return "world"
```

* What is the actual behavior/output?

```plain
main.py:8: error: Invalid signature "def (main.A, Literal['hello']) -> Literal['world']" for "__getattr__"
main.py:12: error: Invalid signature "def (main.A, Literal['hello']) -> Literal['world']" for "__getattribute__"
main.py:12: error: Argument 1 of "__getattribute__" is incompatible with supertype "object"; supertype defines the argument type as "str"
Found 3 errors in 1 file (checked 1 source file)
```

* What is the behavior/output you expect?

```plain
Success: no issues found in 1 source file
```

* What are the versions of mypy and Python you are using? Mypy:mypy 0.770+dev.5336f272c2ac625bd40a79ab3dbbe3d548f147b8, Python:3.8.0
  Do you see the same issue after installing mypy from Git master? yes

**Comment by JukkaL** [comment]
Yeah, this seems like a potential thing to support. I'd assume many real use cases would use an overloaded method to support multiple literals, potentially with a final overload item with signature `(str) -> Any` to handle cases not covered explicitly.

**Comment by bbatliner** [comment]
For my use case, this is about 95% of the way there, _except_ that I can't `@overload` on `__getattr__` or `__getattribute__` successfully.

Here's a working example demonstrating dynamically setting attribute names at runtime, with type hints provided by generics:

```python
A = TypeVar("A", bound=str)
B = TypeVar("B", bound=str)

class Foo(Generic[A, B]):
    @overload
    def __init__(self: "Foo[Literal['a'], Literal['b']]") -> None: ...
    @overload
    def __init__(self: "Foo[A, Literal['b']]", *, a_attr_name: A) -> None: ...
    @overload
    def __init__(self: "Foo[Literal['a'], B]", *, b_attr_name: B) -> None: ...
    @overload
    def __init__(self: "Foo[A, B]", *, a_attr_name: A, b_attr_name: B) -> None: ...
    def __init__(self: "Foo[A, B]", *, a_attr_name: A = cast(A, "a"), b_attr_name: B = cast(B, "b")) -> None:
        setattr(self, str(a_attr_name), True)
        setattr(self, str(b_attr_name), 5)

    @overload
    def fn(self, attr: A) -> bool: ...
    @overload
    def fn(self, attr: B) -> int: ...
    def fn(self, attr: str) -> Any:
        pass

f = Foo()
reveal_type(f.fn("a"))  # Revealed type is "builtins.bool"
reveal_type(f.fn("b"))  # Revealed type is "builtins.int"
```

I say this is 95% of the way there, because if you swap `def fn` for `def __getattr__`, then _voilà_, except mypy doesn't understand the overloaded signatures.

**Comment by lmazuel** [comment]
Just wanted to add that it makes the clean implementation of PEP562 very challenging, as mypy won't recognize the overloads, and claim that `Module "mymodule" has no attribute "Bar"` as well.

**Comment by dchevell** [comment]
> Yeah, this seems like a potential thing to support. I'd assume many real use cases would use an overloaded method to support multiple literals, potentially with a final overload item with signature `(str) -> Any` to handle cases not covered explicitly.

I think it makes more sense without that final overload. `(str) -> Any` would type checkers that all attribute access is valid, so `myobject.thisdoesntexist` would pass type checking.

FWIW I was writing a library where I needed to repetitively define the same attributes on a few different objects. I encoded the names in a `Literal`, constructed everything based on instrospecting the Literal object, then defined `__getattr__` like so:

```
def __getattr__(self, name: ValidNames) -> SomeCustomObject:
        return object.__getattribute__(self, name)
```
In VSCode, pyright accepts this and gives me nice autocompletion for all the names in `ValidNames`, whilst, still highlighting invalid attribute access.
